### PR TITLE
539 explain subnet name

### PR DESCRIPTION
**Pull request by ShiriMoran**

### CODE REVIEW

**Review by zivnevo — CHANGES_REQUESTED** [review]
Need to update flags help and `docs/vpcanalyzer_explain.md`

File: docs/vpcanalyzer_explain.md [hunk 1]
```diff
@@ -20,8 +20,8 @@ vpcanalyzer explain [flags]
 ### Options
 
 ```
-      --src string         source endpoint for explanation; can be specified as a VSI name/CRN or an internal/external IP-address/CIDR;
-                           VSI name can be specified as <vsi-name> or  <vpc-name>/<vsi-name>
+      --src string         source endpoint for explanation; can be specified as a VSI/subnet name/CRN or an internal/external IP-address/CIDR;
+                           VSI/subnet name can be specified as <vsi-name/subnet-name> <vpc-name>/<vsi-name/subnet-name>
       --dst string         destination endpoint for explanation; can be specified as a VSI name/CRN or an internal/external IP-address/CIDR;
```
Comment (zivnevo) [inline docs/vpcanalyzer_explain.md:25]: Need an update here too

File: docs/vpcanalyzer_explain.md [hunk 2]
```diff
@@ -20,8 +20,8 @@ vpcanalyzer explain [flags]
 ### Options
 
 ```
-      --src string         source endpoint for explanation; can be specified as a VSI name/CRN or an internal/external IP-address/CIDR;
-                           VSI name can be specified as <vsi-name> or  <vpc-name>/<vsi-name>
+      --src string         source endpoint for explanation; can be specified as a VSI/subnet name/CRN or an internal/external IP-address/CIDR;
+                           VSI/subnet name can be specified as <vsi-name/subnet-name> <vpc-name>/<vsi-name/subnet-name>
```
Comment (zivnevo) [inline docs/vpcanalyzer_explain.md:24]:
```suggestion
      --src string         source endpoint for explanation; can be specified as a VSI/subnet name/CRN or an internal/external IP-address/CIDR;
                           VSI/subnet name can be specified as <vsi-name/subnet-name> or as <vpc-name>/<vsi-name/subnet-name>
```

File: cmd/analyzer/subcmds/explain.go [hunk 3]
```diff
@@ -27,8 +27,8 @@ const (
 	dstMaxPortFlag = "dst-max-port"
 	detailFlag     = "detail"
 
-	srcDstUsage = "endpoint; can be specified as a VSI name/CRN or an internal/external IP-address/CIDR;\n" +
-		"VSI name can be specified as <vsi-name> or  <vpc-name>/<vsi-name>"
+	srcDstUsage = "endpoint; can be specified as a VSI/subnet name/CRN or an internal/external IP-address/CIDR;\n" +
+		"VSI/subnet name can be specified as <vsi-name/subnet-name> or  <vpc-name>/<vsi-name/subnet-name>"
```
Comment (zivnevo) [inline cmd/analyzer/subcmds/explain.go:31]:
```suggestion
	srcDstUsage = "endpoint; can be specified as a VSI/subnet name/CRN or an internal/external IP-address/CIDR;\n" +
		"VSI/subnet name can be specified as <vsi-name/subnet-name> or as <vpc-name>/<vsi-name/subnet-name>"
```

File: pkg/ibmvpc/examples/out/explain_out/externalToSubnet_all_vpcs_explain_detail.txt [hunk 4]
```diff
@@ -0,0 +1,65 @@
+Explaining connectivity from 161.26.0.0 to subnet3-ky within test-vpc1-ky
+Interpreted source: 161.26.0.0 (external)
+Interpreted destination: vsi3a-ky[10.240.30.5], vsi3b-ky[10.240.30.4], db-endpoint-gateway-ky[10.240.30.6]
```
Comment (zivnevo) [inline pkg/ibmvpc/examples/out/explain_out/externalToSubnet_all_vpcs_explain_detail.txt:3]: Do we want to change that to `Interpreted source(s)` and `Interpreted destination(s)`?

File: pkg/vpcmodel/explainabilityInput.go [hunk 5]
```diff
@@ -306,8 +307,14 @@ func (c *VPCConfig) getNodesFromInputString(cidrOrName string) (nodes []Node,
 	if endpoint != nil {
 		return endpoint, noErr, nil
 	}
+	// 2. cidrOrName references subnet
+	subnetEndpoints := c.getNodesOfSubnet(cidrOrName)
+	if subnetEndpoints != nil {
```
Comment (zivnevo) [inline pkg/vpcmodel/explainabilityInput.go:312]: Is the case where a subnet has no VSIs handled well?

**Review by zivnevo — CHANGES_REQUESTED** [review]

File: pkg/ibmvpc/examples/out/explain_out/externalToSubnet_all_vpcs_explain_detail.txt (hunk 4 above)
Comment (ShiriMoran) [reply pkg/ibmvpc/examples/out/explain_out/externalToSubnet_all_vpcs_explain_detail.txt:3]: I don't think so. The upper header references the original input which is the subnet. Its likely the user knows that the input refers to a subnet (even if we do not explicitly add the term "subnet")

File: pkg/vpcmodel/explainabilityInput.go (hunk 5 above)
Comment (ShiriMoran) [reply pkg/vpcmodel/explainabilityInput.go:312]: depends how you define well ;-) but I guess the answer is no
The error message will be too general. I'll fix it.

File: docs/vpcanalyzer_explain.md (hunk 1 above)
Comment (zivnevo) [reply docs/vpcanalyzer_explain.md:25]: not resolved (flag help for `--dst`)

File: docs/vpcanalyzer_explain.md (hunk 2 above)
Comment (zivnevo) [reply docs/vpcanalyzer_explain.md:24]: not resolved

File: pkg/ibmvpc/examples/out/explain_out/externalToSubnet_all_vpcs_explain_detail.txt (hunk 4 above)
Comment (zivnevo) [reply pkg/ibmvpc/examples/out/explain_out/externalToSubnet_all_vpcs_explain_detail.txt:3]: It's not about what the used known or doesn't know. It's a language thing: singular vs. plural.
The title says `Interpreted destination`, but then 3 different destinations are listed.
Just a little bit weird, but can stay as-is if you want to.

File: pkg/vpcmodel/explainabilityInput.go (hunk 5 above)
Comment (ShiriMoran) [reply pkg/vpcmodel/explainabilityInput.go:316]: fixed, test added

File: docs/vpcanalyzer_explain.md (hunk 1 above)
Comment (ShiriMoran) [reply docs/vpcanalyzer_explain.md:25]: sorry. fixed now

File: pkg/ibmvpc/examples/out/explain_out/externalToSubnet_all_vpcs_explain_detail.txt (hunk 4 above)
Comment (ShiriMoran) [reply pkg/ibmvpc/examples/out/explain_out/externalToSubnet_all_vpcs_explain_detail.txt]: fixed

**Pull request merged**